packages/preset-umi/src/commands/generators/utils.ts

```diff
@@ -194,3 +214,125 @@ export function promptsExitWhenCancel<T extends string = string>(
 export function trim(s?: string) {
   return s?.trim() || '';
 }
+
+interface IFile {
+  from: string;
+  fromFallback: string;
+  to: string;
+  exts?: string[];
+}
+
+export enum ETempDir {
+  Page = 'page',
+  Component = 'component',
+}
+
+export async function processGenerateFiles({
+  filesMap,
+  baseDir,
+  presetArgs = {},
+  templateVars,
+}: {
+  filesMap: IFile[];
+  baseDir: string;
+  presetArgs?: {
+    fallback?: boolean;
+  };
+  templateVars: Record<string, any>;
+}) {
+  const { fallback } = presetArgs;
+  const ignorePattern = /^\.DS_Store$/;
+
+  const choosePath = ({ from, fromFallback, exts = [] }: IFile) => {
+    if (fallback) {
+      return fromFallback;
+    }
+
+    if (existsSync(from)) {
+      if (
+        statSync(from).isDirectory() &&
+        readdirSync(from).filter((name) => !ignorePattern.test(name)).length ===
+          0
+      ) {
+        return fromFallback;
+      }
+
+      return from;
+    }
+
+    for (const ext of exts) {
+      const fullPath = from + ext;
+      if (existsSync(fullPath) && statSync(fullPath).isFile()) {
+        return fullPath;
+      }
+    }
+
+    return fromFallback;
+  };
+
+  const extReg = /.(\.\w+)\.tpl$|.(\.\w+)$/;
+  const names = [];
```

Comment:
```suggestion
  const names: string[] = [];
```

推断不出来的变量初始化，要明确指明类型。